go/signedexchange/structuredheader/parser.go

```diff
@@ -0,0 +1,259 @@
+// Package structuredheader parses Structured Headers for HTTP
+// (https://httpwg.org/http-extensions/draft-ietf-httpbis-header-structure.html).
+package structuredheader
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+// http://httpwg.org/http-extensions/draft-ietf-httpbis-header-structure.html#identifier
+type Identifier string
+
+// http://httpwg.org/http-extensions/draft-ietf-httpbis-header-structure.html#param
+type ParameterisedList []*ParameterisedIdentifier
+type ParameterisedIdentifier struct {
+	Label  Identifier
+	Params Parameters
+}
+
+// Parameter represents a set of parameters in a Parameterised Identifier.
+// The interface{} value is one of these:
+//
+//	int64      for Numbers
+//	string     for Strings
+//	Identifier for Identifiers
+//	[]byte     for Byte Sequences
+//	nil        for parameters with no value
+type Parameters map[Identifier]interface{}
+
+type parser struct {
+	input string
+}
+
+func (p *parser) discardLeadingOWS() {
+	p.input = strings.TrimLeft(p.input, " \t")
+}
+
+func (p *parser) isEmpty() bool {
+	return len(p.input) == 0
+}
+
+func (p *parser) getChar() byte {
+	c := p.input[0]
+	p.input = p.input[1:]
+	return c
+}
+
+func (p *parser) getString(n int) string {
+	s := p.input[:n]
+	p.input = p.input[n:]
+	return s
+}
+
+func (p *parser) consumeChar(c byte) bool {
+	if len(p.input) > 0 && p.input[0] == c {
+		p.input = p.input[1:]
+		return true
+	}
+	return false
+}
+
+// http://httpwg.org/http-extensions/draft-ietf-httpbis-header-structure.html#parse-param-list
+func ParseParameterisedList(input string) (ParameterisedList, error) {
+	p := &parser{input}
+	// In the spec, this is done in the Step 1 of the top-level parsing algorithm.
+	// http://httpwg.org/http-extensions/draft-ietf-httpbis-header-structure.html#text-parse
+	p.discardLeadingOWS()
+
+	var items ParameterisedList
+	for !p.isEmpty() {
+		item, err := p.parseParameterisedIdentifier()
+		if err != nil {
+			return nil, err
+		}
+		items = append(items, item)
+		p.discardLeadingOWS()
+		if p.isEmpty() {
+			return items, nil
+		}
+		if !p.consumeChar(',') {
+			return nil, fmt.Errorf("COMMA expacted, got '%c'", input[0])
```

Comment:
nits: error message should start with small caps in convention (there are a lot of messages don't follow this rule in this project though...)

https://github.com/golang/go/wiki/Errors

How about starting with `structuredheader:`?

Reply:
Done.